### ISSUE TITLE
Directional derivative works different when is called from different place.

### ISSUE THREAD

**Issue by szymag** [sympy/sympy#12969]
Directional derivative is defined in two places. We can use them as a method in `vector` class (is also overloaded):
```python
from sympy.vector import CoordSys3D, directional_derivative, Del
a = CoordSys3D('a')
v = a.i + a.j
delop = Del()
(v & delop)(a.x)
(v & delop)(a.x*v)
```
and the output is:
```python
1
a.i  + a.j
```
But when we use function `directional_derivative` in `sympy/vector/functions.py` we can't calculate directional derivative from vector field as previously, only from scalar field:
```python
directional_derivative(a.x, v)
directional_derivative(v, a.j) 
```
The output:
```python
1
ValueError: Invalid multiplication
```

**Comment by szymag** [comment]
ping @Upabjojr

**Comment by Upabjojr** [comment]
The exception is raised by the `gradient`: you cannot compute the gradient of a vector field.

`(v & delop)(a.x*v)`

with this example you are trying to perform a directional derivative of a vector:

https://en.wikipedia.org/wiki/Directional_derivative#Derivatives_of_vector_valued_functions_of_vectors

I suggest to create a `DirectionalDerivative` class and

- fix the `directional_derivative` function so that it works on vectors (directional derivative of vectors), by moving [this code](https://github.com/sympy/sympy/blob/1a50e9d65eebea3e88041cd3dcb20b24e6a03dcb/sympy/vector/vector.py#L111) to the `directional_derivative` function.
- [This if-clause](https://github.com/sympy/sympy/blob/1a50e9d65eebea3e88041cd3dcb20b24e6a03dcb/sympy/vector/vector.py#L110) should return an instance of the `DirectionalDerivative` class.

**Comment by Upabjojr** [comment]
Otherwise, the `gradient` could be generalized to support vector fields:

https://en.wikipedia.org/wiki/Gradient#Gradient_of_a_vector

**Comment by Upabjojr** [comment]
I have thought a bit more, we basically need to contemplate two cases for the directional derivative:

- you know the directional vector (but not the scalar)
- you know the directional vector and the scalar (or expression) to derive

Both cases should support unevaluated expressions. We could either have two classes (say `DirectionalDerivative` and `DirectionalDerivativeApplied`) or a single class with variable number of parameters. I would prefer a single class.

**Comment by szymag** [comment]
I have already started working on unevaluated expression. 
I also preferred solution with one class.

**Comment by Upabjojr** [comment]
> I also preferred solution with one class.

There is a small problem: if you only have the direction, the class should be callable, otherwise not. Maybe it's better to have two classes.

**Comment by faze-geek** [comment]
Can be closed ref #12991
Results as expected on master -
```
>>> from sympy.vector import CoordSys3D, directional_derivative, Del
>>> a = CoordSys3D('a')
>>> v = a.i + a.j
>>> delop = Del()
>>> (v & delop)(a.x)
1
>>> (v & delop)(a.x*v)
a.i + a.j
>>> directional_derivative(a.x, v)
1
>>> directional_derivative(v, a.j)
0
>>> directional_derivative(a.x*v, v)
a.i + a.j
```
